### ISSUE TITLE
gitlab-ci: external connections issues

### ISSUE THREAD

**Issue by avtikhon** [tarantool/tarantool-qa#60]
Tarantool version:
All

OS version:
All

Bug description:
1. OSX 15 only on mac4
https://gitlab.com/tarantool/tarantool/-/jobs/835453933
```
$ /bin/bash -c "git fetch -p && git checkout -f ${CI_COMMIT_SHORT_SHA}"
fatal: «https://gitlab.com/tarantool/tarantool.git/» недоступно: Could not resolve host: gitlab.com
```

2. On Linux for CentOS *
https://gitlab.com/tarantool/tarantool/-/jobs/835799341
```
Extra Packages for Enterprise Linux 8 - x86_64  7.4 kB/s | 3.4 kB     00:00    
Failed to download metadata for repo 'epel'
Error: Failed to download metadata for repo 'epel'
Extra Packages for Enterprise Linux 8 - x86_64  7.4 kB/s | 3.4 kB     00:00    
Failed to download metadata for repo 'epel'
Error: Failed to download metadata for repo 'epel'
make: *** [/pack//rpm.mk:162: package] Error 1
```

3. On Linux for Fedora *
https://github.com/tarantool/tarantool/runs/1933471340
```
Fedora Modular 29 - x86_64 - Updates            0.0  B/s |   0  B     00:29    
Failed to synchronize cache for repo 'updates-modular'
Error: Failed to synchronize cache for repo 'updates-modular'
make: *** [/pack//rpm.mk:162: package] Error 1
```

**Comment by avtikhon** [comment]
Used strace to collect information on the issue, got the following log:
[centos-7.zip](https://github.com/tarantool/tarantool-qa/files/6075108/centos-7.1.zip)
Found the following:
Found at line 10953:
```
237   openat(AT_FDCWD, "/lib64/libnss_dns.so.2", O_RDONLY|O_CLOEXEC) = 11
237   read(11, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0p\23\0\0\0\0\0\0"..., 832) = 832
237   fstat(11, {st_mode=S_IFREG|0755, st_size=32904, ...}) = 0
237   mmap(NULL, 28912, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 11, 0 <unfinished ...>
236   <... select resumed>)             = 0 (Timeout)
````
also on line 11222
```
237   connect(11, {sa_family=AF_INET6, sin6_port=htons(443), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "2605:bc80:3010:600:dead:beef:cafe:fed9", &sin6_addr), sin6_scope_id=0}, 28 <unfinished ...>
236   rt_sigaction(SIGPIPE, NULL,  <unfinished ...>
237   <... connect resumed>)            = -1 ENETUNREACH (Network is unreachable)
```
issue in loop starting from line 11264:
```
236   connect(11, {sa_family=AF_INET, sin_port=htons(443), sin_addr=inet_addr("152.19.134.142")}, 16) = -1 EINPROGRESS (Operation now in progress)
236   poll([{fd=11, events=POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout)
236   rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f820bccc030}, NULL, 8) = 0
236   rt_sigaction(SIGPIPE, NULL, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f820bccc030}, 8) = 0
236   rt_sigaction(SIGPIPE, {sa_handler=SIG_IGN, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7f820bccc030}, NULL, 8) = 0
236   poll([{fd=11, events=POLLOUT|POLLWRNORM}], 1, 0) = 0 (Timeout)
```

**Comment by Totktonada** [comment]
@avtikhon raised the question on GitHub support. We'll wait for an answer for a while.

**Comment by Totktonada** [comment]
I see the same ENETUNREACH from my local machine with only link-local IPv6 for an external IPv6 address (verified with ping6).

However it is still unclear whether it is the root problem.

**Comment by Totktonada** [comment]
Can you collect `/var/log/dnf.log` and `/var/log/dnf.librepo.log` for the failing case as well?

**Comment by Totktonada** [comment]
Got something like this locally, but the logs does not reveal much: just 30 seconds timeout for the metalink repository:

> 2021-03-04T17:30:40Z DEBUG Cannot download 'https://mirrors.fedoraproject.org/metalink?repo=updates-released-f29&arch=x86_64': Cannot prepare internal mirrorlist: Curl error (28): Timeout was reached for https://mirrors.fedoraproject.org/metalink?repo=updates-released-f29&arch=x86_64 [Connection timed out after 30000 milliseconds].

[dnf-logs-fedora-29.zip](https://github.com/tarantool/tarantool-qa/files/6085636/dnf-logs-fedora-29.zip)

**Comment by Totktonada** [comment]
NB: 'Could not resolve host: gitlab.com' is the different problem (and I don't know whether it still exists).

----

When `dnf makecache` is successful I see two requests to `https://mirrors.fedoraproject.org/metalink?repo=updates-released-f29`, in `dnf.librepo.log` (where the same URL is stated as the effective one). They are clarified with the following [log messages](https://github.com/rpm-software-management/librepo/blob/928c2d9c66f0d10477b8c9d1192785b88e2b1155/librepo/handle.c#L1214-L1232):

* Only fetching mirrorlist/metalink
* Downloading/Locating yum repo

I guess those two requests are made against `mirrors.fedoraproject.org` per se, not a mirror.

After those steps the `repodata/repomd.xml` file and other files are downloaded. Each such request is prepended with 'Selecting mirror for: \<filename\>' message.

My failed case is the following. The 'Only fetching mirrorlist/metalink' request is successful, but the 'Downloading/Locating yum repo' is not. I'll try to list `librepo` changes that looks relevant.

First, the table of librepo versions we have in the build images (collected with `dnf makecache && dnf info librepo`):

| ${DIST} | packpack/packpack:fedora-${DIST} librepo installed | fedora:${DIST} librepo installed | fedora:${DIST} librepo available |
| --- | --- | --- | --- |
| 28 | 1.8.1-7.fc28 | 1.8.1-7.fc28 | 1.8.1-7.fc28 |
| 29 | 1.9.5-1.fc29 | 1.9.4-1.fc29 | **1.10.5-1.fc29** |
| 30 | 1.11.1-1.fc30 | 1.11.1-1.fc30 | 1.11.1-1.cf30 |
| 31 | 1.11.3-1.fc31 | 1.11.3-1.fc31 | **1.12.1-1.fc31** |
| 32 | 1.11.1-4.fc32 | 1.12.1-1.fc32 | **1.12.1-1.fc32** |
| 33 | 1.12.11.fc33 |  1.12.1-1.fc33| 1.12.1-1.fc33 |
| 34 | N/A | 1.12.1-2.fc34 | **1.13.0-1.fc34** |
| 35 (rawhide) | N/A | 1.12.1-2.fc34 | **1.13.0-1.fc35** |

I looked at the `librepo` changes that look relevent to the temporary connectivity problems. They're the following:

| Change | Released as | Comment |
| --- | --- | --- |
| https://github.com/rpm-software-management/librepo/pull/158 | 1.10.5 | It is unclear for me, whether it affects retrying of 'Downloading/Locating yum repo' requests / relevant to metalink request retrying |
| https://github.com/rpm-software-management/librepo/pull/167 | 1.10.6 | Unclear in this sense too |
| https://github.com/rpm-software-management/librepo/pull/169 | 1.11.0 | Seems quite relevant to the metalink request retrying. It also adds logging of IP addresses that is nice for debugging. |
| https://github.com/rpm-software-management/librepo/pull/189 | 1.12.0 | I guess not relevant, because we have only metalink URL enabled, but mentioned, because it is around the same topic |
| https://github.com/rpm-software-management/librepo/pull/217 | 1.13.0 | I hope this is another problem, but who knows |

Nothing points to zchunk, so I skipped several commits relevant to zchunk. But I'll highlight this [one](https://github.com/rpm-software-management/librepo/commit/76c833e80680fbc8b00158033669cda43b44c2ef), because it explicitly mentions the yandex.ru mirror as problematic (I don't know whether it is still so). This change is released in 1.10.2.

So I would say that librepo of at least 1.11.0 version is desirable for fetching reliability. But it is unavalilable on Fedora 28/29 and already installed on Fedora 30+.

We really need some failure statistics with logs (`/var/log/dnf.log` and `/var/log/dnf.librepo.log`) for each fail if we want to solve this problem somehow. I propose to start collecting it.

Maybe dropping support of old Fedora distros is the solution. Maybe not. I would not make any decisions without understanding of the problem or at least good statistics (not just 'it seems, it fails rarely after attempt to fix X', but something that is collected for a long time in CI).

----

Redarding the [attempt to fix it](https://github.com/packpack/packpack/pull/136) using clearing the dnf cache. It is not bad thing per se (but should be done during image build, not in the packpack tool). However I doubt it actually changes something around the problem with synchronizing a repository metainfo. At least, from what I see, old cached metalink mirrors are not used.

**Comment by avtikhon** [comment]
According to [documentation](https://readthedocs.org/projects/dnf/downloads/pdf/latest/), there are the following types available cache clean:

- **_dnf clean dbcache_** Removes cache files generated from the repository metadata. This forces DNF to regenerate the cache files the next time it is run.
- _**dnf clean expire-cache**_ Marks the repository metadata expired. DNF will re-validate the cache for each repository the next time it is used.
- _**dnf clean metadata**_ Removes repository metadata. Those are the files which DNF uses to determine the remote availability of packages. Using this option will make DNF download all the metadata the next time it is run.
- _**dnf clean packages**_ Removes any cached packages from the system.
- _**dnf clean all**_ Does all of the above.

Running each of the types locally and on Github Actions got:
| Type | packages | metadata | dbcache | expire-cache | all |
|---|---|---|---|---|---|
| Local output 1st line | 0 files removed | Cache was expired | 11 files removed | Cache was expired | 36 files removed |
| Local output 2nd line || 36 files removed || 0 files removed ||

Running much more workflows than in previous found that clean up didn't completely fixed the issue, also found that in Fedora 29 was deprecated option [failovermethod](https://forums.fedoraforum.org/showthread.php?320926-failovemethod-option). Decided to run dnf update to be sure that packages are up-to-date and found that many of them updated even DNF itself. After it even much more runs began to pass always.
 
 | Package | Version |
 |---|---|
| dnf | 4.2.5-5.fc29 |
| dnf-data | 4.2.5-5.fc29 |
| dnf-plugins-core | 4.0.7-1.fc29 |
| dnf-utils | 4.0.7-1.fc29 |
| dnf-yum | 4.2.5-5.fc29 |
